util/activity-report.js

```diff
@@ -83,24 +47,140 @@ async function queryAllActivity(
   };
 }
 
+/* 
+generates time periods to query past activities:
+[
+  [ 2022-07-13T22:00:00.000Z, 2022-07-20T22:00:00.000Z ],
+  [ 2022-07-06T22:00:00.000Z, 2022-07-13T22:00:00.000Z ],
+] 
+*/
+function getTimeIntervals(
+  apiStartDate,
+  apiEndDate,
+  periods, 
+) {
+  const startDate = new Date(apiStartDate);
+  const endDate = new Date(apiEndDate);
+  
+  if ([startDate, endDate].some((date) => !(date instanceof Date && !isNaN(date))))
+    throw new Error(`date has unsupported format`);
+  if (startDate >= endDate)
+    throw new Error('start day should be earlier than end date');
+  if (isNaN(periods) || periods < 1 || periods > 20)
+    throw new Error(`Illegal moving average 'periods' parameter`);
+
+  const intervalLength = endDate - startDate;
+  const timeIntervals = [];
+  for (let i = 0; i < periods; i++) {
+    const dateTo = new Date(endDate - intervalLength * i);
+    const dateFrom = new Date(dateTo - intervalLength);
+    timeIntervals.push([dateFrom, dateTo]);
+  }
+  return {
+    timeIntervals,
+    intervalLength,
+  };
+}
+
+function getMovingAverage(type = 'simple', data = []) {
+  let reducer;
+  switch (type) {
+    case 'weighted':
+      // sum of the data array indices
+      const sum = [...data.keys()].reduce((p, c) => p + c + 1, 0);
+      reducer = (p, c, i) => p + (c * (i + 1) / sum);
+      break;
+    case 'simple':
+    default:
+      reducer = (p, c) => p + c / data.length;
```

Comment:
this part might benefit from more comments about the code (e.g. example output for a common input), plus more descriptive variable names (instead of `p`, `c`, and `i`)

Reply:
also, for `sum` you've got an `O(n)` method of calculating this, but there is an `O(1)` calculation, use `n*(n+1)/2)` to calculate this.